### PR TITLE
INJECTOR: Add type declarations to properties

### PR DESCRIPTION
**Pull request by sjokkateer**
And also missing method return types

Signed-off-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description
This PR adds type declarations to the `Injector::class` properties as well as update some of the missing method return types.

### CODE REVIEW

File: src/Injector.php [hunk 1]
```diff
@@ -26,20 +26,13 @@
...
-
-    /** @var ContainerInterface */
-    protected $container;
-
-    /** @var DependencyResolverInterface */
-    protected $resolver;
-
-    /** @var ConfigInterface */
-    protected $config;
+    protected DefinitionInterface $definition;
```
Comment (Ocramius) [inline src/Injector.php]: This type addition is a BC break for inheritance scenarios 🤔

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Injector.php (hunk 1 above)
Comment (sjokkateer) [reply src/Injector.php]: Would you like me to add back the docblock or is there a better way to account for the BC break?

Comment (Ocramius) [reply src/Injector.php]: Back to comment is okay

Comment (sjokkateer) [reply src/Injector.php]: On a side note, is there a way for me to detect BC breaks like those up front?

File: src/Injector.php [hunk 2]
```diff
@@ -28,18 +28,12 @@ class Injector implements InjectorInterface
...
-
-    /** @var ConfigInterface */
-    protected $config;
+    protected ContainerInterface $container;
+    protected DependencyResolverInterface $resolver;
+    protected ConfigInterface $config;
 
     /** @var string[] */
-    protected $instantiationStack = [];
+    protected array $instantiationStack = [];
```
Comment (samsonasik) [inline src/Injector.php:36]: this is BC break, only `private` property that allowed the type to be changed when class is not final

Comment (sjokkateer) [reply src/Injector.php:36]: I see, thanks for the insight. Does that mean the other protected properties I modified should be reverted back to using docblocks as well?

Comment (samsonasik) [reply src/Injector.php:36]: yes

File: src/Injector.php [hunk 3]
```diff
@@ -113,11 +107,10 @@ public function canCreate(string $name): bool
      *
      * @param string $name Class name or service alias
      * @param array  $parameters Constructor parameters, keyed by the parameter name.
-     * @return object|null
      * @throws ClassNotFoundException
      * @throws RuntimeException
      */
-    public function create(string $name, array $parameters = [])
+    public function create(string $name, array $parameters = []): ?object
```
Comment (samsonasik) [inline src/Injector.php:113]: this is also BC break when extended class override it, ref https://3v4l.org/QTiTt

File: src/Injector.php [hunk 4]
```diff
@@ -145,11 +138,10 @@ public function create(string $name, array $parameters = [])
      *
      * @param string $name The type name to instantiate.
      * @param array  $params Constructor arguments, keyed by the parameter name.
-     * @return object
      * @throws InvalidCallbackException
      * @throws ClassNotFoundException
      */
-    protected function createInstance(string $name, array $params)
+    protected function createInstance(string $name, array $params): object
```
Comment (samsonasik) [inline src/Injector.php:144]: this is also BC break when extended class override it, ref https://3v4l.org/OgOHS

Comment (sjokkateer) [reply src/Injector.php:144]: I see now, thanks again for the insight. 
Sorry for my lack of knowledge and wasting your and @Ocramius time but I think this pull request can be closed/ignored then.

File: src/Injector.php (hunk 1 above)
Comment (Ocramius) [reply src/Injector.php]: I wrote https://github.com/Roave/BackwardCompatibilityCheck, which is somewhat able to detect some of these :-) 

I'd really like to integrate it with our CI: https://github.com/laminas/laminas-ci-matrix-action/issues/60

**Pull request closed without merging**